nemo/collections/multimodal/speech_llm/models/modular_s2s_models.py

```diff
@@ -1139,27 +1158,32 @@ def resample(audio, audio_lens, orig_sample_rate, target_sample_rate):
                 )
                 sliced_text_channel = text_channel[: answer_codec.shape[0]]
 
-            if getattr(self.cfg, 'speech_delay', False):
-                text_len, text_vocab = sliced_text_channel.shape
-                speech_len, speech_vocab = answer_codec.shape
-                assert text_len == speech_len
-                speech_pad_id = self.cfg.data.train_ds.speech_unk_id
-                text_pad_id = self.tokenizer.eos_id
-                answer_codec_padded = torch.full(
-                    (self.cfg.speech_delay, speech_vocab), speech_pad_id, device=answer_codec.device
-                )
-                answer_codec_shifted = torch.cat([answer_codec_padded, answer_codec], dim=0)[:speech_len, :]
-                sliced_text_channel_padded = torch.full(
-                    (self.cfg.speech_delay, text_vocab), text_pad_id, device=sliced_text_channel.device
-                )
-                sliced_text_channel_extended = torch.cat([sliced_text_channel, sliced_text_channel_padded], dim=0)[
-                    :speech_len, :
-                ]
-                combined_channels = torch.cat([sliced_text_channel_extended, answer_codec_shifted], dim=-1)
-                all_channels.append(combined_channels)
+            # import pdb; pdb.set_trace()
+
+            if getattr(self.cfg, 'predict_source_text', False):
+                all_channels.append(torch.cat([sliced_text_channel, answer_codec, sliced_source_text_channel], dim=-1))
```

Comment:
have u verified in predict_source_text==True case the all_channels[0,:,-1] is as expected?

Reply:
![Screenshot 2024-12-19 at 14 11 44](https://github.com/user-attachments/assets/02db593b-3f2b-4f66-a6b1-c9f591afc5ba)
Yes, it roughly looks like this, first and last channels are target and source text, respectively. And I print out some speech channels in between.

Reply:
ah i got the point. so source text is predicted after the finish of user speech
we can test whether it has zero shot streaming ASR ability by setting 1 earlier in the inference pipeline in future